content/algorithms/Isomorphism/isomorphism.md

```diff
@@ -0,0 +1,221 @@
+  name: python3
+---
+
+# **Isomorphism**
+
+```{code-cell} ipython3
+import networkx as nx
+import matplotlib.pyplot as plt
+```
+
+## What's isomorphism? Why is it interesting?
+
++++
+
+As unlabeled graphs can have multiple spatial representations, two graphs are isomorphic if they have the same number of edges, vertices, and same edges connectivity. Let's see an example of two isomorphic graphs, 
+
+```{code-cell} ipython3
+plt.subplot(221)
+G = nx.Graph([(0, 1), (0, 4), (0,2), (1, 3), (1, 5), (2, 3), (2, 6), (3, 7), (4, 5), (5, 7), (6, 7), (6, 4)])
+nx.draw_spring(G, with_labels = True, node_color = "c")
+plt.title("G", fontweight="bold")
+H = nx.Graph([(1, 2), (1, 5), (1,3), (2, 4), (2, 6), (3, 4), (3, 7), (4, 8), (5, 6), (6, 8), (7, 8), (7, 5)])
+plt.subplot(222)
+nx.draw_circular(H, with_labels = True, node_color = "yellow")
+plt.title("H", fontweight="bold")
+
+plt.show()
+```
+
+These graphs' spatial representations are very different yet they are isomorphic.  
+
++++
+
+### **Formal definition**
+
++++
+
+G and H are isomorphic if we can establish a bijection between the vertex sets of G and H. 
+
+$${\displaystyle f\colon N(G)\to N(H)}$$
+such as if $$ $$
+
+<center> $v$  and $ w $ are  adjacent  in G $\iff$ $f(v)$ and $f(w)$ are adjacent in H </center>
+
++++
+
+To formally prove that 2 graphs are isomorphic we need to find the bijection between the vertex set. For the previous example that would be: 
+
+$$f(i) = i+1 \hspace{0.5cm} \forall i \in [0, 7]$$ 
+
++++
+
+For small examples, isomorphism may seem easy. But it isn't a simple problem. For two graphs G and H of n nodes, there are n! bijections function possible. Checking every combination is not a feasible option for bigger graphs. 
+In fact, isomorphism is part of the problems known as NP. This means that we don't know any algorithm that runs in polynomial time.
+
++++
+
+### Applications (TODO: Should I move this to the end?)
+
+- Verification of equivalence of different representations of the design of an electronic circuit and communication networks. 
+- Image recognition.
+- Identification of chemical compounds and proteins.
+- Algorithms for fingerprint, facial and retina matching. 
+-  Clustering Algorithms on social networks.
+
++++
+
+## Isomorphism Algorithms
+**Naive Approach**
+
++++
+
+There are some initial properties that we can check to decide whether it's possible to have an isomorphism
+- G and H have the same amount of nodes and edges 
+- The degree sequence for G and H are the same
+
+These are necessary conditions but don't guarantee that 2 graphs are isomorphic. Let's see a small example:
+
+```{code-cell} ipython3
+plt.subplot(221)
+G = nx.cycle_graph(6)
+nx.draw_circular(G)
+plt.title("G", fontweight="bold")
+plt.subplot(222)
+H = nx.union(nx.cycle_graph(3), nx.cycle_graph(3), rename = ("s","d"))
+nx.draw_circular(H, node_color = "r")
+plt.title("H", fontweight="bold")
+plt.show()
+```
+
+```{code-cell} ipython3
+nx.faster_could_be_isomorphic(G, D)
```

Comment:
Maybe it helps with adding a short comment about what the function `faster_could_be_isomorphic` is checking for? It's not really intuitive from the name of the function.